Fix hardcoded conditional rendering of client secret input field

useresd opened the pull request:
Client Secret field conditional rendering is now depending on the ClientAuthenticatorFactory method supportsSecret(). Which allows the ability to implement the ClientAuthenticator provider that uses a client secret without having the client secret input field disappear from the Admin UI.

Closes #22660

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

useresd committed:
Fix hardcoded conditional rendering of client secret field

Client Secret field conditional rendering is now depending on the ClientAuthenticatorFactory method supportsSecret(). Which allows the ability to implement the ClientAuthenticator provider that uses a client secret without having the client secret input field disappear from the Admin UI.

Closes #22660

jonkoops commented on js/apps/admin-ui/src/clients/credentials/Credentials.tsx:
```diff
@@ -200,10 +204,8 @@ export const Credentials = ({ client, save, refresh }: CredentialsProps) => {
-            clientAuthenticatorType === "client-secret-jwt") && <Divider />}
-          {(clientAuthenticatorType === "client-secret" ||
-            clientAuthenticatorType === "client-secret-jwt") && (
+          {selectedProvider?.supportsSecret && <Divider />}
+          {selectedProvider?.supportsSecret && (
```
Looks like we can probably have just one statement here and render both the `Divider` and the `CardBody`, not sure why it wasn't like that in the first place.

useresd replied:
Okay. I will fix that.

useresd committed:
collapsed a redundant conditional statement

jonkoops reviewed (dismissed):
These changes look good to me, I'd still like to see another of the team members take a look at this, especially the Java code.

ssilvert reviewed (dismissed):
LGTM

jonkoops commented:
Hmmm, it looks like these changes are causing some of the tests to fail. Those will have to fixed first before we can merge this change in. I'll do a rebase just to make sure this is not some kind of flake, but I doubt it is.

jonkoops commented:
Thanks for letting me know @mhajas. I am still curious if the failing Cypress tests have been impacted by this PR, or if the cause was some other work that was merged.

useresd commented:
The test were failing because of this pull request.

I was able to reproduce the same issues that was on cypress, and it was because of my commit.

The changes I've made to the [buildProviderMetadata](https://github.com/keycloak/keycloak/blob/de448920fcd7ab0e5a4cd17bf7ad6b556c01d6cc/services/src/main/java/org/keycloak/services/resources/admin/AuthenticationManagementResource.java#L164) method affected other endpoints.

I changed the casting of the factory from ConfigurableAuthenticatorFactory to ClientAuthenticatorFactory. But this method is used by other resource that are not an implementation ClientAuthenticatorFactories.

so the following endpoints were affected:
- `/admin/realms/${realmName}/authentication/form-providers`
- `/admin/realms/${realmName}/authentication/authenticator-providers`
- `/admin/realms/${realmName}/authentication/form-action-providers`

I fixed the issue. Did the same failing cypress tests manually and everything seems to work fine now.

I will push the commit.

useresd committed:
fixed changes made in PR #22681.

useresd committed:
Merge branch 'main' into issue-22660

jonkoops commented:
Looks like we still have one failing test, but things are already looking better.

useresd commented:
So, any idea on how to fix that failing test? I can see two tests were failing.

jonkoops commented:
Looks like in some cases the tests are now getting a `supportsSecret` where they were not before. I don't think that is a bug, so likely we can just update the expected value.

jonkoops commented:
@ useresd are you still interested in landing this? We have a policy of not keeping PRs out for too long, so let me know if you are still interested.

useresd commented:
Yes I am. I will change the tests expected value so the tests could pass.

ssilvert commented:
@useresd Tests still failing.  Please fix.

ssilvert commented:
@useresd Are you still working on this?

useresd committed:
fixed ClientAuthenticatorProviders test assertions to include supportsSecret in the expected value

useresd committed:
Merge remote-tracking branch 'origin/issue-22660' into issue-22660

useresd commented:
Yes. I fixed the test assertion. The test is now passing

```bash
stsuite/integration-arquillian-tests-other/999.0.0-SNAPSHOT/integration-arquillian-tests-other-999.0.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Keycloak Arquillian Integration TestSuite 999.0.0-SNAPSHOT:
[INFO] 
[INFO] Keycloak Arquillian Integration TestSuite .......... SUCCESS [  0.372 s]
[INFO] Test apps .......................................... SUCCESS [  0.008 s]
[INFO] Test apps distribution ............................. SUCCESS [  0.622 s]
[INFO] Keycloak Authz Tests: Hello World Example .......... SUCCESS [  0.460 s]
[INFO] Keycloak Authz: Servlet Authorization Test ......... SUCCESS [  0.231 s]
[INFO] Keycloak Authz: Simple Servlet App with Policy Enforcer SUCCESS [  0.065 s]
[INFO] integration-arquillian-test-apps-servlets .......... SUCCESS [  0.617 s]
[INFO] Keycloak Test App Profile JEE ...................... SUCCESS [  0.090 s]
[INFO] integration-arquillian-test-apps-cors-parent ....... SUCCESS [  0.010 s]
[INFO] Angular Product Portal JS .......................... SUCCESS [  0.259 s]
[INFO] JAX-RS Database Service Using OAuth Bearer Tokens .. SUCCESS [  0.102 s]
[INFO] spring-boot-adapter-app ............................ SUCCESS [  0.203 s]
[INFO] Fuse Test Applications ............................. SUCCESS [  0.008 s]
[INFO] Customer Portal - Secured in Karaf/Fuse ............ SUCCESS [  0.847 s]
[INFO] CXF JAXRS Example - Secured in Karaf/Fuse .......... SUCCESS [  0.201 s]
[INFO] CXF JAXRS Example - Secured in Karaf/Fuse 7 on Undertow SUCCESS [  0.161 s]
[INFO] CXF JAXWS Example - Secured in Karaf/Fuse .......... SUCCESS [  0.117 s]
[INFO] CXF JAXWS Example - Secured in Karaf/Fuse 7 on Undertow SUCCESS [  0.132 s]
[INFO] Product Portal - Secured in Karaf/Fuse ............. SUCCESS [  0.115 s]
[INFO] Product Portal - Secured in Karaf/Fuse 7 on Undertow SUCCESS [  0.150 s]
[INFO] Camel endpoint example - Secured in Karaf/Fuse ..... SUCCESS [  0.143 s]
[INFO] Camel endpoint example - Secured in Karaf/Fuse 7.0 on Undertow SUCCESS [  0.162 s]
[INFO] Keycloak Fuse Example - Features ................... SUCCESS [  0.078 s]
[INFO] Keycloak Examples - External Config ................ SUCCESS [  0.170 s]
[INFO] Servers ............................................ SUCCESS [  0.010 s]
[INFO] Testsuite Adapter SPI .............................. SUCCESS [  0.007 s]
[INFO] Adapter Undertow SPI (JakartaEE) ................... SUCCESS [  0.355 s]
[INFO] Undertow OIDC Adapter (JakartaEE) .................. SUCCESS [  0.423 s]
[INFO] Undertow SAML Adapter (JakartaEE) .................. SUCCESS [  0.438 s]
[INFO] Auth Server ........................................ SUCCESS [  0.008 s]
[INFO] Auth Server Services ............................... SUCCESS [  0.010 s]
[INFO] Auth Server Services - Testsuite Providers Deployment SUCCESS [  0.060 s]
[INFO] Auth Server Services - Testsuite Providers ......... SUCCESS [  1.381 s]
[INFO] Test Util .......................................... SUCCESS [  0.470 s]
[INFO] Auth Server - Undertow ............................. SUCCESS [  0.640 s]
[INFO] Auth Server - Quarkus .............................. SUCCESS [  8.211 s]
[INFO] App Server ......................................... SUCCESS [  0.005 s]
[INFO] App Server - SPI ................................... SUCCESS [  0.043 s]
[INFO] App Server - JBoss ................................. SUCCESS [  0.007 s]
[INFO] App Server - Wildfly ............................... SUCCESS [ 25.993 s]
[INFO] App Server - Karaf ................................. SUCCESS [  0.005 s]
[INFO] App Server - Tomcat ................................ SUCCESS [  0.006 s]
[INFO] App Server - Undertow .............................. SUCCESS [  0.257 s]
[INFO] App Server - Jetty Parent .......................... SUCCESS [  0.009 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server SUCCESS [  0.008 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server - Infinispan SUCCESS [  0.009 s]
[INFO] Keycloak Arquillian Integration TestSuite - Cache Server - Legacy SUCCESS [  0.009 s]
[INFO] Tests .............................................. SUCCESS [  0.058 s]
[INFO] Base TestSuite ..................................... SUCCESS [ 28.615 s]
[INFO] Other Tests Modules ................................ SUCCESS [  0.006 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:13 min
[INFO] Finished at: 2023-09-27T14:42:15+03:00
[INFO] ------------------------------------------------------------------------

```

useresd committed:
Merge branch 'main' into issue-22660

jonkoops commented:
@useresd there are still failing tests on CI, and it seems that the branch now has conflicts with `main`. Could you resolve them and take a look at fixing the tests?

jonkoops commented:
@useresd are you still interested in landing this PR? If so let us know, otherwise we will close it to prune out our backlog.

useresd commented:
> @useresd are you still interested in landing this PR? If so let us know, otherwise we will close it to prune out our backlog.

Sure. I am interested in landing this PR. I will fix the conflict and make sure the other tests are passing.

useresd committed:
Merge branch 'main' into issue-22660

jonkoops reviewed (approved):
Front-end changes look good to me, @ssilvert could you take a look at the other stuff?

ssilvert reviewed (approved):
LGTM

ssilvert commented:
@useresd Please fix conflicts.

jonkoops commented:
@useresd Are you still interested in landing this? We have a convention of closing PRs that have been out for too long without activity, so we'll likely be closing this one soon.

ImFlog commented:
I have the exact same issue and I am interested in the PR.
If @useresd don't have enough time to help on this I could restart the MR with the same code and fix conflicts.
WDYT ?

useresd commented:
I tried to fix failing tests, but more tests keep failing. Running the tests locally on my machine takes a lot of times and consumes bandwidth. I will try to give it a final shot today.

jonkoops commented:
Closing this in favor of #25776.
